lib/completion-templates.ts

```diff
@@ -5,45 +5,47 @@ export const completionShTemplate = `###-begin-{{app_name}}-completions-###
 # Installation: {{app_path}} {{completion_command}} >> ~/.bashrc
 #    or {{app_path}} {{completion_command}} >> ~/.bash_profile on OSX.
 #
-_{{app_name}}_yargs_completions()
+_{{safe_app_name}}_yargs_completions()
 {
-    local cur_word args type_list
+    if [[ {{app_name_word_check}} ]]; then
+      local cur_word args type_list
 
-    cur_word="\${COMP_WORDS[COMP_CWORD]}"
-    args=("\${COMP_WORDS[@]}")
+      cur_word="\${COMP_WORDS[COMP_CWORD]}"
+      args=("\${COMP_WORDS[@]}")
 
-    # ask yargs to generate completions.
-    type_list=$({{app_path}} --get-yargs-completions "\${args[@]}")
+      # ask yargs to generate completions.
+      type_list=$({{app_path}} --get-yargs-completions "\${args[@]}")
 
-    COMPREPLY=( $(compgen -W "\${type_list}" -- \${cur_word}) )
+      COMPREPLY=( $(compgen -W "\${type_list}" -- \${cur_word}) )
 
-    # if no match was found, fall back to filename completion
-    if [ \${#COMPREPLY[@]} -eq 0 ]; then
-      COMPREPLY=()
+      # if no match was found, fall back to filename completion
+      if [ \${#COMPREPLY[@]} -eq 0 ]; then
+        COMPREPLY=()
+      fi
     fi
 
     return 0
 }
-complete -o bashdefault -o default -F _{{app_name}}_yargs_completions {{app_name}}
+complete -o bashdefault -o default -F _{{safe_app_name}}_yargs_completions {{app_name_first_word}}
 ###-end-{{app_name}}-completions-###
 `;
 
-export const completionZshTemplate = `#compdef {{app_name}}
+export const completionZshTemplate = `#compdef "{{safe_app_name}}"
 ###-begin-{{app_name}}-completions-###
 #
 # yargs command completion script
 #
 # Installation: {{app_path}} {{completion_command}} >> ~/.zshrc
 #    or {{app_path}} {{completion_command}} >> ~/.zprofile on OSX.
 #
-_{{app_name}}_yargs_completions()
+_{{safe_app_name}}_yargs_completions()
 {
   local reply
   local si=$IFS
   IFS=$'\n' reply=($(COMP_CWORD="$((CURRENT-1))" COMP_LINE="$BUFFER" COMP_POINT="$CURSOR" {{app_path}} --get-yargs-completions "\${words[@]}"))
   IFS=$si
   _describe 'values' reply
 }
-compdef _{{app_name}}_yargs_completions {{app_name}}
+compdef _{{safe_app_name}}_yargs_completions "{{zsh_app_name}}"
```

Comment:
`compdef` makes it easy, as it supports multiple words right our of the box. We just have to replace any spaces with a `=`. I don't think the quotes I added are required, but they also can't hurt.

Reply:
`yarn=foo` does not install a completion handler for the subcommand `foo` of command `yarn`, it installs a completion handler for command `yarn` _and_ command `foo`.

Reply:
https://github.com/zsh-users/zsh-completions/blob/master/zsh-completions-howto.org#copying-completions-from-another-command

Reply:
With code in `index.mjs`:
```
  .scriptName('yarn foo')
```

and script in `package.json`:
```
  "scripts": {
    "foo": "node index.mjs"
  },
```

Setup:
```
% yarn foo completion > comp.sh
% source comp.sh

```

Get completion as intended here:
```console
 % yarn foo --
--boolean  --string  --                                                                        
--help               -- Show help                                                              
--version            -- Show version number     
```

But also here:
```console
% yarn --
--boolean  --string  --                                                                        
--help               -- Show help                                                              
--version            -- Show version number     
```

And here:
```console
% foo --
--boolean  --string  --                                                                        
--help               -- Show help                                                              
--version            -- Show version number     
```

Reply:
(The above comments all assuming zsh.)

Reply:
Ah, I hadn't noticed this. I'll look into adding a check for the whole command in the competition handler in following week.